### PR TITLE
Collect access denied errors on describe

### LINKED ISSUE #347: `describe` command should fetch all the information possible and display failures in the end
**Describe the problem/challenge you have**
Sometimes when trying to get more information about a Bundle an error can occur while talking with the registry or the user might not have access to some particular Image. Today, when `imgpkg` fails to retrieve information about some image it displays the error message and stops processing. Nevertheless, `imgpkg` could still provide information about the rest of the Images that are part of the Bundle. 

**Describe the solution you'd like**

When running the command `imgpkg describe` the command should run until the end without falling and in the end provide information about the failures

```
Bundle SHA: sha256:3cd8f0d917a92fafeef2d44eb63195d5351c35cf46b8b7448cb126c7d68c6291

Images:
  - Image: localhost:5000/test@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
    Type: Image
    Origin: index.docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
    Annotations:
      kbld.carvel.dev/id: index.docker.io/dkalinin/k8s-simple-app:latest
      kbld.carvel.dev/origins: - resolved:
    tag: latest
    url: index.docker.io/dkalinin/k8s-simple-app:latest


  - Image: localhost:5000/test@sha256:510482df49db421542ae10142e8ce99572b23ef72675c37e92e4a12b541a3f6a
    Type: Internal

  - Image: localhost:5000/test@sha256:ecc10da4f480b1ba263d96197151d76bc787104cbcf7f1b1f42b14340b58bae7
    Type: Signature
    Error: Access Denied

Unable to describe the full bundle, please review the previous output to understand the error
Failed
```

When the error happens while reading an Image from ImagesLock, `imgpkg` should assume that the type of the Image is `Image` and should provide the following output

```
    Image: localhost:5000/test@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
    Type: Image
    Origin: index.docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
    Error: Image Not Found
```

**Anything else you would like to add:**
For signatures/sboms/internal Images we should ignore 404 errors since they might not be present, nevertheless, if we get a different error it should be logged.
This behavior should apply to all the output types.

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help work on this issue.

### PR DESCRIPTION
**Pull request by joaopapereira**
Fixes #347

### CODE REVIEW

File: pkg/imgpkg/bundle/images_refs.go [hunk 1]
```diff
@@ -60,6 +61,20 @@ func NewImageRefWithType(imgRef lockconfig.ImageRef, imageType ImageType) ImageR
 	return ImageRef{ImageRef: imgRef, IsBundle: &isBundle, Copiable: copiable, ImageType: imageType}
 }
 
+// NewImageRefWithTypeAndError Constructs a new ImageRef based on the ImageType that failed to fetch
+func NewImageRefWithTypeAndError(imgRef lockconfig.ImageRef, imageType ImageType, err string) ImageRef {
+	copiable := true
```
Comment (neil-hickey) [inline pkg/imgpkg/bundle/images_refs.go:66]: Suggestion:

copiable := (imageType == InternalImage)

File: pkg/imgpkg/signature/fetch_signatures_test.go [hunk 2]
```diff
@@ -44,6 +45,34 @@ func TestSignatureRetriever_Signatures(t *testing.T) {
...
+
+		var args []lockconfig.ImageRef
+		args = append(args, lockconfig.ImageRef{Image: "registry.io/img@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0"})
+		args = append(args, lockconfig.ImageRef{Image: "registry.io/img1@sha256:6716afd7a68262a37d3f67681ed9dedf3b882938ad777f268f44d68894531f7a"})
+		args = append(args, lockconfig.ImageRef{Image: "registry.io/img2@sha256:56cb33b3b4bc45509c5ff7513ddc6ed78764f9ad5165cc32826e04da49d5462b"})
+		signatures, err := subject.FetchForImageRefs(args)
+		require.Error(t, err)
+		errs, ok := err.(*signature.FetchError)
+		require.True(t, ok, "Unexpected error found %+v, while expecting a FetchError", err)
+		require.Len(t, errs.AllErrors, 1)
```
Comment (neil-hickey) [inline pkg/imgpkg/signature/fetch_signatures_test.go]: is it worth this have >1 errors in it? Just to validate that when more than one access denied error occurs, that they are all shown/returned?

File: pkg/imgpkg/bundle/images_refs.go [hunk 3]
```diff
@@ -60,6 +61,20 @@ func NewImageRefWithType(imgRef lockconfig.ImageRef, imageType ImageType) ImageR
 	return ImageRef{ImageRef: imgRef, IsBundle: &isBundle, Copiable: copiable, ImageType: imageType}
 }
 
+// NewImageRefWithTypeAndError Constructs a new ImageRef based on the ImageType that failed to fetch
+func NewImageRefWithTypeAndError(imgRef lockconfig.ImageRef, imageType ImageType, err string) ImageRef {
```
Comment (neil-hickey) [inline pkg/imgpkg/bundle/images_refs.go:59]: Suggestion:

copiable := (imageType == InternalImage)

**Review by neil-hickey — COMMENTED** [review]
LGTM, just one question about the testing scope.

File: pkg/imgpkg/signature/fetch_signatures_test.go (hunk 2 above)
Comment (joaopapereira) [reply pkg/imgpkg/signature/fetch_signatures_test.go]: The main reason for this check is the one after where we do
`require.EqualError(t, errs.AllErrors[0], "access denied")`
to make sure we do not panic or something.

Are you asking if we should have a test that contains more than 1 error?

Comment (neil-hickey) [reply pkg/imgpkg/signature/fetch_signatures_test.go]: > Are you asking if we should have a test that contains more than 1 error?

This

**Review by cppforlife — APPROVED** [review]

**Pull request merged**